proxylib/binarymemcached/parser.go

```diff
@@ -0,0 +1,328 @@
+// Copyright 2018 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//
+// Accompanying file `headerparser.policy` contains an example policy
+// for this protocol. Install it with:
+// $ cilium policy import proxylib/testparsers/headerparser.policy
+//
+
+package binarymemcached
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"github.com/cilium/cilium/pkg/envoy/cilium"
+	"github.com/cilium/cilium/proxylib/proxylib"
+
+	log "github.com/sirupsen/logrus"
+)
+
+// BinaryMemcacheRule matches against memcached requests
+type BinaryMemcacheRule struct {
+	// opCode group name
+	opCode string
+
+	keyExact  string
+	keyPrefix string
+	keyRegex  string
+
+	// allowed opcodes
+	opCodes []byte
+	// compiled regex
+	regex *regexp.Regexp
+}
+
+// Matches returns true if the BinaryMemcacheRule matches
+func (rule *BinaryMemcacheRule) Matches(data interface{}) bool {
+	log.Infof("binarymemcache checking rule %v", *rule)
+
+	packet := data.([]byte)
+	key, err := getMemcacheKey(packet)
+
+	if err != nil {
+		log.Errorf("error retrieving binary memcached key: %s", err)
+	}
+
+	opcode := packet[1]
+
+	if rule.keyExact == "" && rule.keyPrefix == "" && rule.keyRegex == "" {
+		return rule.matchOpcode(opcode)
+	}
+
+	if rule.keyExact != "" && rule.keyExact == key {
+		return rule.matchOpcode(opcode)
+	}
+	log.Infof("binary memcached exact key %s does not match %s", rule.keyExact, key)
+
```

Comment:
If the rule gives an exact key to match, and it fails to match we should return 'false'? I.e., security rules should be formulated so that each part of the rule should match. If you want to allow either a key match or a keyPrefix match, you should have separate rules for these cases instead of integrating them to the same rule.

Reply:
So on a single parsed rule only one of the three key match alternatives should be allowed. With that this function would be much simpler, returning `false` on the first mismatch.